### PR TITLE
integrate documents generation from Chocolate Doom

### PR DESCRIPTION
**Pull request by rfomin**
Fixes #603 
It currently only generates `CMDLINE.md`, I want to know what you think of this approach.
After some experimentation, I think it should be integrated into the build system, it's nice to document all options in code without thinking about generation.

`docgen` can output plain text, which is arguably more useful than Markdown, but on Windows it has wrong newlines. Alternatively, we can convert Markdown to HTML using `pandoc` like Choco.

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
> `docgen` can output plain text, which is arguably more useful than Markdown, but on Windows it has wrong newlines. 

I guess we could replace `sys.stdout.buffer.write(buf)` with `print(buf.replace('\n', os.linesep))` in `stdout()`.

> Alternatively, we can convert Markdown to HTML using `pandoc` like Choco.

Sure, pandoc is powerful, but also carries a lot of dependencies and is currently not available in MSYS2.

**Comment by rfomin** [comment]
> I guess we could replace `sys.stdout.buffer.write(buf)` with `print(buf.replace('\n', os.linesep))` in `stdout()`.

It didn't work. I just replaced it with `print()`:
```diff
--- a/man/docgen
+++ b/man/docgen
@@ -482,7 +482,8 @@ def print_template(template_file, substs, content):
                 line = line.replace("@content", content)
                 for k,v in substs.items():
                     line = line.replace(k,v)
-                stdout(line.rstrip().encode('UTF-8') + b'\n')
+                print(line.rstrip())
+                # stdout(line.rstrip().encode('UTF-8') + b'\n')
     finally:
         f.close()
 
@@ -501,7 +502,8 @@ def wiki_output(targets, _, template):
     read_wikipages()
 
     for t in targets:
-        stdout(t.wiki_output().encode('UTF-8') + b'\n')
+        print(st.wiki_output())
+        # stdout(t.wiki_output().encode('UTF-8') + b'\n')
 
 def markdown_output(targets, substs, template_file):
     content = ""
```
Do you know why `stdout()` is needed?

**Comment by rfomin** [comment]
I think we don't need config and woof-setup documentation? I also haven't figured out how bash-completion works yet.

**Comment by rfomin** [comment]
Should we do something with README.md? Convert it to text or html or just remove it?

**Comment by rfomin** [comment]
@fabiangreffrath Should I continue this? These new comments don't always fit well with the MBF codebase, but I like it.

**Comment by fabiangreffrath** [comment]
I'll review this tomorrow or on Monday. Looks good so far. 👍

File: CMakeLists.txt [hunk 1]
```diff
@@ -149,6 +149,9 @@ find_package(SDL2_net)
 # FluidSynth backend
 find_package(FluidSynth)
 
+# Python 3
+find_package(Python3 COMPONENTS Interpreter)
```
Comment (fabiangreffrath) [inline CMakeLists.txt:153]: So, python will now be a hard dependency?

File: man/CMDLINE.template.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Command line parameters
```
Comment (fabiangreffrath) [inline man/CMDLINE.template.md:1]: Do we need both files, `man/CMDLINE.template` and `man/CMDLINE.template.md`?

File: man/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (fabiangreffrath) [inline man/CMakeLists.txt:1]: Blank line?

File: man/woof.template [hunk 4]
```diff
@@ -0,0 +1,51 @@
+.TH @PROJECT_SHORTNAME@ 6
+.SH NAME
+@PROJECT_SHORTNAME@ \- Woof! is a continuation of Lee Killough's Doom source port MBF targeted
+at modern systems
+.SH SYNOPSIS
+.B woof
+[\fIOPTIONS\fR]
+.SH DESCRIPTION
+.PP
+MBF stands for "Marine's Best Friend" and is regarded by many as the successor
```
Comment (fabiangreffrath) [inline man/woof.template:10]: Could you adjust the wording to the latest version provided in `README.md`?

File: src/d_deh.c [hunk 5]
```diff
@@ -2576,6 +2576,13 @@ void deh_procCheat(DEHFILE *fpin, FILE* fpout, char *line) // done
 
   // [FG] ignore cheats in dehacked files
   boolean deh_apply_cheats = true;
+
+  //!
+  // @category mod
+  //
+  // Ignore cheats in dehacked files.
```
Comment (fabiangreffrath) [inline src/d_deh.c:2583]: I appreciate your respectful editing, but please remove redundant comments, even if preceeded by my initials. 😉

File: src/d_main.c [hunk 6]
```diff
@@ -869,7 +897,15 @@ void IdentifyVersion (void)
     screenshotdir = M_StringDuplicate(basesavegame);
   }
 
-  // [FG] set screenshot path to -shotdir parm or fall back to -save parm or current dir
+  // [FG] set screenshot path to -shotdir parm or fall back to -save parm or
```
Comment (fabiangreffrath) [inline src/d_main.c:900]: Redundant comment.

**Comment by fabiangreffrath** [comment]
> Should we do something with README.md? Convert it to text or html or just remove it?

Not sure what you mean? README.md is the project's homepage, no way we remove it. But you mean at install phase, right?

**Comment by rfomin** [comment]
> But you mean at install phase, right?

Yes. I like Markdown, but I'm not sure about other users. It's a little thing not sure if anyone reads it.

File: man/CMDLINE.template.md (hunk 2 above)
Comment (rfomin) [reply man/CMDLINE.template.md:1]: I'm thinking about adding CMDLINE.md to the wiki. But it seems it's not possible to automate it.

File: src/d_main.c [hunk 7]
```diff
@@ -1474,6 +1528,13 @@ static void D_AutoloadPWadDehDir()
 static void D_ProcessDehInWad(int i, boolean in_iwad)
 {
   // [FG] avoid loading DEHACKED lumps embedded into WAD files
+
```
Comment (fabiangreffrath) [inline src/d_main.c:1531]: Here is another redundant comment.

**Comment by fabiangreffrath** [comment]
> Do you know why `stdout()` is needed?

No, I have no idea. Maybe to support Python2 as well for which `print()` isn't a function?

File: man/CMDLINE.template.md (hunk 2 above)
Comment (fabiangreffrath) [reply man/CMDLINE.template.md:1]: So we can remove one of both?

Comment (rfomin) [reply man/CMDLINE.template.md:1]: But now I am generating CMDLINE.md. I think we can manually copy its contents to the wiki every release.

**Comment by fabiangreffrath** [comment]
Could you please add the `-mb` parameter?

**Comment by rfomin** [comment]
> Could you please add the `-mb` parameter?

`docgen` crash on `z_zone.c`, I guess too many macros.

**Comment by fabiangreffrath** [comment]
> `docgen` crash on `z_zone.c`, I guess too many macros.

Sigh, we need to tidy up that mess at some time...

**Pull request merged**